subsys/fs/ext2/ext2_diskops.c

```diff
@@ -0,0 +1,906 @@
+		return 3;
+	}
+	return -1;
+}
+
+int64_t ext2_inode_remove_blocks(struct ext2_inode *inode, uint32_t first)
+{
+	uint32_t start;
+	int max_lvl;
+	int64_t ret, removed = 0;
+	uint32_t offsets[4];
+
+	max_lvl = get_level_offsets(inode->i_fs, first, offsets);
+
+	if (all_zero(offsets, max_lvl)) {
+		/* We remove also the first block because all blocks referenced from it will be
+		 * deleted.
+		 */
+		start = offsets[0];
+	} else {
+		/* There will be some blocks referenced from first affected block hence we can't
+		 * remove it.
+		 */
+		__ASSERT(inode->i_block[offsets[0]] != 0,
+				"Block that references some other must be nonzero");
+
+		start = offsets[0] + 1;
+		ret = delete_blocks(inode->i_fs, inode->i_block[offsets[0]],
+				block0_level(offsets[0]), &offsets[1]);
+		if (ret < 0) {
+			return ret;
+		}
+		removed += ret;
+	}
+
+	for (uint32_t i = start; i < EXT2_INODE_BLOCKS; i++) {
+		if (inode->i_block[i] == 0) {
+			continue;
+		}
+		ret = delete_blocks(inode->i_fs, inode->i_block[i], block0_level(i),
+				zero_offsets);
+		if (ret < 0) {
+			return ret;
+		}
+		removed += ret;
+		inode->i_block[i] = 0;
+	}
+	return removed;
+}
+
+static inline uint32_t get_ngroups(struct ext2_data *fs)
+{
+	uint32_t ngroups =
+		EXT2_DATA_SBLOCK(fs)->s_blocks_count / EXT2_DATA_SBLOCK(fs)->s_blocks_per_group;
+
+	if (EXT2_DATA_SBLOCK(fs)->s_blocks_count % EXT2_DATA_SBLOCK(fs)->s_blocks_per_group != 0) {
+		/* there is one more group if the last group is incomplete */
+		ngroups += 1;
+	}
+	return ngroups;
+}
+
+int ext2_fetch_block_group(struct ext2_data *fs, uint32_t group)
+{
+	if (fs->bgroup == NULL) {
+		fs->bgroup = ext2_heap_alloc(sizeof(struct ext2_bgroup));
+		if (fs->bgroup == NULL) {
+			return -ENOMEM;
+		}
+
+		LOG_DBG("Allocated new bgroup (%p) for fs (%p)", fs->bgroup, fs);
+	}
+
+	struct ext2_bgroup *bg = fs->bgroup;
+
+	/* Check if block group is cached */
+	if (bg->block && group == bg->num) {
+		return 0;
+	}
+
+	uint32_t ngroups = get_ngroups(fs);
+
+	LOG_DBG("ngroups:%d", ngroups);
+	LOG_DBG("cur_group:%d fetch_group:%d", bg->num, group);
+
+	if (group > ngroups) {
+		return -ERANGE;
+	}
+
+	uint32_t groups_per_block = fs->block_size / sizeof(struct ext2_disk_bgroup);
+	uint32_t block = group / groups_per_block;
+	uint32_t offset = group % groups_per_block;
+
+	/* fetch block group block if not cached */
+	if ((bg->block == NULL) || (block != bg->block->num)) {
+		uint32_t global_block = EXT2_DATA_SBLOCK(fs)->s_first_data_block + 1 + block;
+
+		ext2_drop_block(bg->block);
+		bg->block = ext2_get_block(fs, global_block);
+		if (bg->block == NULL) {
+			return -ENOENT;
+		}
+	}
+
+	/* Invalidate previously fetched blocks */
+	ext2_drop_block(bg->inode_table);
+	ext2_drop_block(bg->inode_bitmap);
+	ext2_drop_block(bg->block_bitmap);
+	bg->inode_table = bg->inode_bitmap = bg->block_bitmap = NULL;
+
+	bg->fs = fs;
+	bg->num = group;
+	bg->num_in_block = offset;
+
+	struct ext2_disk_bgroup *dg = current_disk_bgroup(bg);
```

Comment:
Seems that this is only used by debug log. Isn't this wasted call in all other instances?